s3datarepository/iam.go

```diff
@@ -177,8 +177,40 @@ func (s *S3Repository) deletePolicy(ctx context.Context, id string) error {
 		return ErrCode("failed to get ARN for policy "+policyName, err)
 	}
 
-	// TODO: check if policy is used anywhere before deleting
+	// check if this policy is used anywhere and detach it before deleting
 
+	log.Debugf("listing roles with policy %s", policyArn)
+
+	// find out what entities the policy is attached to
+	// TODO: right now we only check roles, but may need to handle groups/users eventually
+	entitiesOut, err := s.IAM.ListEntitiesForPolicyWithContext(ctx, &iam.ListEntitiesForPolicyInput{
+		EntityFilter:      aws.String("Role"),
+		PathPrefix:        aws.String(s.IAMPathPrefix),
+		PolicyArn:         aws.String(policyArn),
+		PolicyUsageFilter: aws.String("PermissionsPolicy"),
+	})
+	if err != nil {
+		log.Warnf("failed to list entities for policy "+policyArn, err)
+	}
+
+	if entitiesOut != nil {
+		log.Debug(entitiesOut.PolicyRoles)
+		if len(entitiesOut.PolicyRoles) == 0 {
+			log.Debugf("policy %s is not attached to any roles", policyName)
+		}
```

Comment:
this log statement conditional seems excessive... 😆 

`log.Debugf("policy %s is attached to %d roles", policyName, len(entitiesOut.PolicyRoles))`